src/vmm/src/seccomp.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::io::Read;
+use std::sync::Arc;
+
+use bincode::{DefaultOptions, Error as BincodeError, Options};
+
+/// Each BPF instruction is 8 bytes long and 4 byte aligned.
+/// This alignment needs to be satisfied in order for a BPF code to be accepted
+/// by the syscalls. Using u64 here is is safe as it has same size and even bigger alignment.
+pub type BpfInstruction = u64;
+
+/// Program made up of a sequence of BPF instructions.
+pub type BpfProgram = Vec<BpfInstruction>;
+
+/// Reference to program made up of a sequence of BPF instructions.
+pub type BpfProgramRef<'a> = &'a [BpfInstruction];
+
+/// Type that associates a thread category to a BPF program.
+pub type BpfThreadMap = HashMap<String, Arc<BpfProgram>>;
+
+/// Retrieve empty seccomp filters.
+pub fn get_empty_filters() -> BpfThreadMap {
+    let mut map = BpfThreadMap::new();
+    map.insert("vmm".to_string(), Arc::new(vec![]));
+    map.insert("api".to_string(), Arc::new(vec![]));
+    map.insert("vcpu".to_string(), Arc::new(vec![]));
+    map
+}
+
+/// Binary filter deserialization errors.
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum DeserializationError {
+    /// Bincode deserialization failed: {0}
+    Bincode(BincodeError),
+}
+
+/// Deserialize binary with bpf filters
+pub fn deserialize_binary<R: Read>(
+    reader: R,
+    bytes_limit: Option<u64>,
+) -> Result<BpfThreadMap, DeserializationError> {
+    let result = match bytes_limit {
+        Some(limit) => DefaultOptions::new()
+            .with_fixint_encoding()
+            .allow_trailing_bytes()
+            .with_limit(limit)
+            .deserialize_from::<R, HashMap<String, BpfProgram>>(reader),
+        // No limit is the default.
+        None => bincode::deserialize_from::<R, HashMap<String, BpfProgram>>(reader),
+    }
+    .map_err(DeserializationError::Bincode)?;
+
+    Ok(result
+        .into_iter()
+        .map(|(k, v)| {
+            println!("bpf len: {}", v.len());
+            (k.to_lowercase(), Arc::new(v))
+        })
+        .collect())
+}
+
+/// Filter installation errors.
+#[derive(Debug, PartialEq, Eq, thiserror::Error, displaydoc::Display)]
+pub enum InstallationError {
+    /// Filter length exceeds the maximum size of {BPF_MAX_LEN:} instructions
+    FilterTooLarge,
+    /// prctl` syscall failed with error code: {0}
+    Prctl(i32),
+}
+
+/// The maximum seccomp-BPF program length allowed by the linux kernel.
+pub const BPF_MAX_LEN: usize = 4096;
```

Comment:
ah, so ever worrying about whether our seccomp rules were larger than a page was moot anyway cuz the kernel would've rejected then :rofl:

Reply:
ah, makes sense